### ISSUE TITLE
Mistake in CoSENTLoss documentation

### ISSUE THREAD

**Issue by pashok3d** [huggingface/sentence-transformers#3368]
In the CoSENTLoss implementation, documentation and the docstring states:
> `loss = logsum(1 + exp(s(i,j) - s(k,l)) + …)`, where `(i,j)` and `(k,l)` are any of the input pairs in the batch such that the expected similarity of `(i,j)` is greater than `(k,l)`.

However, the actual code masks on
https://github.com/UKPLab/sentence-transformers/blob/dd8a8d7d4d5660e7b2b69e8a1ece8202417c0069/sentence_transformers/losses/CoSENTLoss.py#L89
meaning it keeps pair-differences only when the true label of the first pair is less than the second. And computes
https://github.com/UKPLab/sentence-transformers/blob/dd8a8d7d4d5660e7b2b69e8a1ece8202417c0069/sentence_transformers/losses/CoSENTLoss.py#L86
(i.e. always s_i - s_j). As a result, the loss sums exp(s_i - s_j) over all (i,j) where y_i < y_j, not y_i > y_j as the docstring claims.

So it should be:
`loss = logsum(1 + exp(s(i,j) - s(k,l)) + …)`, where `(i,j)` and `(k,l)` are any of the input pairs in the batch such that the expected similarity of `(i,j)` is **smaller** than `(k,l)`.

**Comment by pashok3d** [comment]
@tomaarsen Hi! Yes, indeed, it is related to that comment. IMO there was a confusion in https://github.com/UKPLab/sentence-transformers/issues/2437#issuecomment-2304623160 and #3230 PR actually changed correct description to incorrect one. 
The confusion arises because the image in https://github.com/UKPLab/sentence-transformers/issues/2437#issuecomment-2304623160 shows AnglE loss function that uses $s(u,v)$ for similarity and $\Delta\theta_{ij}$ for angle **difference**, whereas the formulation of loss function in documentation uses only **similarities**. Similarity (cosine) is **inversely** related to the angle difference, therefore it should be:

For CoSENT it should be:
``loss = logsum(1+exp(s(k,l)-s(i,j))+exp...)``, where ``(i,j)`` and ``(k,l)`` are any of the input pairs in the batch such that the expected similarity of ``(i,j)`` is greater than ``(k,l)``.

But for AnglE it should be:
``loss = logsum(1+exp(Δθ_ij-Δθ_kl)+exp...)``, where ``(i,j)`` and ``(k,l)`` are any of the input pairs in the batch such that the expected similarity of ``(i,j)`` is greater than ``(k,l)``, which means that the expected angle difference of ``(i,j)`` is smaller than that of ``(k,l)``.

**Comment by johneckberg** [comment]
@pashok3d Thank you for noticing this! I will take a look first thing tomorrow morning.

**Comment by johneckberg** [comment]
@pashok3d, @tomaarsen Sorry for the delay! I've been taking a look again at the implementations and the source paper, [AoE: Angle-optimized Embeddings for Semantic Textual Similarity](https://aclanthology.org/2024.acl-long.101.pdf). I believe I've figured out what the issue is and would appreciate any thoughts/input. (TL;DR: potential typo in the paper) 

Unfortunately,  angle distance as described in the paper is not inversely proportional to cosine similarity. It exists to provide a better learning signal than normal cosine similarity, as there is very small gradient when "normal" cosine similarity is close to the top or the bottom of the waveform.

Here are the three functions of interest (in LaTeX which does not render in the mobile app):

## The CoSENT loss function from the initial [blog post](https://kexue.fm/archives/8847)

$$\mathcal{L}_{\cos} =  \log \left( 1 + \sum _{\text{sim(i,j)} > \text{sim(k,l)}} e^{\lambda(\cos(u_k, u_l) - \cos(u_i, u_j))} \right)$$


## The CoSENT loss function from the AnglE paper (pre-print only)

$$\mathcal{L}_{cos} = log \left[ 1 + \sum _{s(X_i, X_j) > s(X_m, X_n)} e^{\frac{\cos(X_m, X_n) - \cos(X_i, X_j)}{\tau}} \right]$$

## The AnglE loss function from the AnglE paper (pre-print & ACL)

$$\mathcal{L}_{angle} = log \left[ 1 + \sum _{s(X_i, X_j) > s(X_m, X_n)} e^{\frac{\Delta\theta _{ij} - \Delta\theta _{mn}}{r}} \right]$$

**Notice that from CoSENT loss to AnglE loss, the order of $s(m,n) - s(i,j)$ in the summation is changed to $s(i,j) - s(m,n)$. However, the actual implementation of the paper does not reflect this.**

Each of these losses use: ``y_pred = y_pred[:, None] - y_pred[None, :]`` to create a batch size by batch size matrix that contains the difference in similarity between each possible pairs of input pairs. Let's denote this difference as $score(k,l) - score(i,j)$. 

Note the following lines from each loss function; each function creates **the same mask** for the matrix of differences.

## The Original [CoSENT loss](https://github.com/bojone/CoSENT/blob/124c368efc8a4b179469be99cb6e62e1f2949d39/cosent.py#L79) implementation

 ```
y_true = K.cast(y_true[:, None] < y_true[None, :], K.floatx())
```

## The Angle Optimized [cosine_loss](https://github.com/SeanLee97/AnglE/blob/main/angle_emb/angle.py#L107) (CoSENT loss) Implementation

 ```
 y_true = (y_true[:, None] < y_true[None, :]).float()
```

## The Angle Optimized [angle_loss](https://github.com/SeanLee97/AnglE/blob/main/angle_emb/angle.py#L114) (AnglE loss) Implementation

 ```
 y_true = (y_true[:, None] < y_true[None, :]).float()
```

A value [(k,l),(i,j)] in the mask matrix is 1 if the ground truth similarity score for pair k,l is less than the ground truth similarity score for pair i,j. Then,  each of these losses use``y_pred = (y_pred - (1 - y_true) * 1e12)`` , which uses the term ``(1 - y_true)`` to create a new mask where 1's correspond to all irrelevant pairs (where the ground truth score for pair k,l is greater or equal than the ground truth score for pair i,j). Multiplying this new mask by $1*e^{12}$ and subtracting it from the irrelevant pairs in the scores matrix makes them become negligible in the ``logsumexp()`` . 

This means that for **all** of the aforementioned implementations of AnglE or CoSENT, ``.view(-1)``, leaves us with a flattened score vector made up of $s(k,l) - s(i,j)$ for $scores(i,j) > scores(k,l)$ **not**  $s(i,j) - s(k,l)$ for $scores(i,j) > scores(k,l)$.

In the previous issue [#2437 ](https://github.com/UKPLab/sentence-transformers/issues/2437#issuecomment-2304623160) discussing this, I assumed that I had made a mistake when writing the documentation. This lead me to quickly change the documentation to reflect the screenshot included in the issue, without actually doing any further due diligence. 

That was a mistake on my part.  @pashok3d is correct in saying there is an error in the documentation, however, there seems to also be a slight difference between the equation given in the AnglE paper and the actual code for the loss function. I think a small typo in the paper is likely.

**Comment by tomaarsen** [comment]
Wow, that is very tricky. Thank you for looking into this. Pinging @SeanLee97 who might benefit from knowing that there's a potential typo in his paper. 

- Tom Aarsen

**Comment by johneckberg** [comment]
Circling back on this issue; it's been a bit, but I think it's important to get this documentation right to help prevent user confusion. 

When I did the original PR #3230 to fix the CoSENTLoss docs, I really should've just changed the AnglE docs, as the documentation error noted in issue #2437 was in reference to the formula in the AnglE paper, not CoSENT. I think the best way forward is to just get the docs fixed so they align with their respective papers.

Here's what I'm thinking:

1. The CoSENT loss documentation should be reverted back to what is was before PR #3230 to match the formula that's in both the [2022 blog post](https://kexue.fm/archives/8847) and [2024 paper](https://penghao-bdsc.github.io/papers/CoSENT_TASLP2024.pdf) on CoSENT loss. I think it makes sense as well to update the doc string for CoSENT loss to reference the 2024 paper published by the IEEE, not the 2022 blog post.

3. The AnglE Loss documentation should be updated to match the formula described in the [2024 AoE paper](https://aclanthology.org/2024.acl-long.101/), regardless of if there's a typo in the formula or not. Similarly to CoSENT loss, I think it also makes sense to update the AnglE doc string to point to the version published by the ACL, instead of the [pre-print](https://arxiv.org/abs/2309.12871v1) on Arxiv it currently references.

This would ensure that the documentation for both losses is correct according to the original research papers.

Let me know if this plan makes sense. If so, I can put together a PR (or two) for this. Thoughts @tomaarsen @SeanLee97?

**Comment by tomaarsen** [comment]
Seems good from my perspective! Feel free to open the PR(s).

- Tom Aarsen

**Comment by tomaarsen** [comment]
This should be solved by #3496 now, closing.

- Tom Aarsen
